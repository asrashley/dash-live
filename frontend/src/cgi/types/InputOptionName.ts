export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}