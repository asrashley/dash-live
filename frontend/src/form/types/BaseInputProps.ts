import type { ReadonlySignal } from "@preact/signals";
import { StaticInputProps } from "./StaticInputProps";

export type BaseInputProps = {
  name: string;
  id: string;
  type: StaticInputProps["type"];
  className: string | ReadonlySignal<string>;
  title: string;
  placeholder?: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

