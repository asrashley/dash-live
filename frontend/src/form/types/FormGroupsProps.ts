import { InputFormGroup } from "./InputFormGroup";
import { FormRowMode } from "./FormRowMode";
import { SetValueFunc } from "./SetValueFunc";
import { InputProps } from "./InputProps";

export interface FormGroupsProps extends Pick<InputProps, "data" | "disabledFields"> {
  groups: InputFormGroup[];
  expand?: string;
  mode: FormRowMode;
  layout?: number[];
  setValue: SetValueFunc;
}
