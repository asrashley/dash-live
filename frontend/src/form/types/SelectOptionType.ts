export interface SelectOptionType {
    name?: string;
    title: string;
    value: string;
}
