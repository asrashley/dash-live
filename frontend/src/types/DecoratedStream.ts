import { Stream } from "./Stream";
import { StreamTrack } from "./StreamTrack";


export interface DecoratedStream extends Stream {
    tracks: StreamTrack[];
}
