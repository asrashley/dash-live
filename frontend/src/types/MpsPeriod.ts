import { MpsTrack } from './MpsTrack';

export type MpsPeriod = {
    duration: string;
    ordering: number;
    parent: number;
    pid: string;
    pk: number | string;
    start: string;
    stream: number;
    tracks: MpsTrack[];
    new?: boolean;
};
