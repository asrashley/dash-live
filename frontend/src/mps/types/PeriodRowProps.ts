import { MpsPeriod } from "../../types/MpsPeriod";

export interface PeriodRowProps {
  period: MpsPeriod;
}
