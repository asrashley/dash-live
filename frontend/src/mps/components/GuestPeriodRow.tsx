import { Fragment } from "preact";
import { useCallback, useContext } from "preact/hooks";

import { AppStateContext } from "../../appState";
import { TimeDeltaInput } from "../../components/TimeDeltaInput";
import { AllStreamsContext } from "../../hooks/useAllStreams";
import { PeriodRowProps } from "../types/PeriodRowProps";
import { TrackSelectionButton } from "./TrackSelectionButton";

function doNothing() {
}

export function GuestPeriodRow({ period }: PeriodRowProps) {
  const { ordering, pid, pk, start, duration } = period;
  const { streamsMap } = useContext(AllStreamsContext);
  const { dialog } = useContext(AppStateContext);
  const stream = streamsMap.value.get(`${period.stream}`);
  const selectTracks = useCallback(() => {
    dialog.value = {
      backdrop: true,
      trackPicker: {
        pk,
        pid,
        guest: true,
        stream,
      },
    };
  }, [dialog, pid, pk, stream]);

  return <Fragment>
    <div className="col period-ordering">{ordering}</div>
    <div className="col period-id">{pid}</div>
    <div className="col period-stream">{stream?.title}</div>
    <div className="col period-start">
      <TimeDeltaInput
        value={start}
        name={`start_${pk}`}
        onChange={doNothing}
        disabled />
    </div>
    <div class="col period-duration">
      <TimeDeltaInput
        value={duration}
        name={`duration_${pk}`}
        onChange={doNothing}
        disabled />
    </div>
    <TrackSelectionButton period={period} stream={stream} selectTracks={selectTracks} />
  </Fragment>;
}
