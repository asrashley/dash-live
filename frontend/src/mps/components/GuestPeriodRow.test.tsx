import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { fireEvent } from "@testing-library/preact";
import { signal } from "@preact/signals";

import { renderWithProviders } from "../../test/renderWithProviders";
import { MpsPeriod } from "../../types/MpsPeriod";
import {
  AllStreamsContext,
  decorateAllStreams,
  UseAllStreamsHook,
} from "../../hooks/useAllStreams";
import { DecoratedStream } from "../../types/DecoratedStream";
import { GuestPeriodRow } from "./GuestPeriodRow";

import { streams } from "../../test/fixtures/streams.json";
import { model } from "../../test/fixtures/multi-period-streams/demo.json";

describe("GuestPeriodRow component", () => {
  const allStreams = signal<DecoratedStream[]>([]);
  const loaded = signal<boolean>(false);
  const streamsMap = signal<Map<string, DecoratedStream>>(new Map());
  const error = signal<string | null>(null);
  const allStreamsHook: UseAllStreamsHook = {
    allStreams,
    loaded,
    streamsMap,
    error,
  };

  beforeEach(() => {
    loaded.value = true;
    error.value = null;
    allStreams.value = decorateAllStreams(streams);
    const sMap = new Map<string, DecoratedStream>();
    allStreams.value.forEach((item) => {
      sMap.set(`${item.pk}`, item);
    });
    streamsMap.value = sMap;
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test.each(model.periods)(
    "matches snapshot for period $pid",
    async (item: MpsPeriod) => {
      const { asFragment, findByText } = renderWithProviders(
        <AllStreamsContext.Provider value={allStreamsHook}>
          <GuestPeriodRow index={2} item={item} className="guest-period-row" />
        </AllStreamsContext.Provider>
      );
      const decStream = streamsMap.value.get(`${item.stream}`);
      expect(decStream).toBeDefined();
      await findByText(decStream.title);
      expect(asFragment()).toMatchSnapshot();
    }
  );

  test("opens track view dialog", () => {
    const period = model.periods[0];
    const { getBySelector, state } = renderWithProviders(
      <AllStreamsContext.Provider value={allStreamsHook}>
        <GuestPeriodRow index={2} item={period} className="guest-period-row" />
      </AllStreamsContext.Provider>
    );
    const elt = getBySelector(".period-tracks > .btn") as HTMLButtonElement;
    fireEvent.click(elt);
    const decStream = streamsMap.value.get(`${period.stream}`);
    expect(decStream).toBeDefined();
    expect(state.dialog.value).toEqual({
      backdrop: true,
      trackPicker: {
        pk: period.pk,
        pid: period.pid,
        guest: true,
        stream: decStream,
      },
    });
  });
});
