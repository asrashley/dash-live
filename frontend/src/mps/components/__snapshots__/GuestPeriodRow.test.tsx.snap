// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GuestPeriodRow component > matches snapshot for period 'p1' 1`] = `
<DocumentFragment>
  <div
    class="col period-ordering"
  >
    2
  </div>
  <div
    class="col period-id"
  >
    p1
  </div>
  <div
    class="col period-stream"
  >
    Big Buck Bunny
  </div>
  <div
    class="col period-start"
  >
    <input
      class="form-control "
      disabled=""
      min="00:00:00"
      name="start_1"
      step="1"
      type="time"
    />
  </div>
  <div
    class="col period-duration"
  >
    <input
      class="form-control "
      disabled=""
      min="00:00:00"
      name="duration_1"
      step="1"
      type="time"
    />
  </div>
  <div
    class="col period-tracks"
  >
    <button
      aria-disabled="false"
      class="btn btn-sm m-1 btn-warning"
    >
      0/3 tracks
    </button>
  </div>
</DocumentFragment>
`;

exports[`GuestPeriodRow component > matches snapshot for period 'p2' 1`] = `
<DocumentFragment>
  <div
    class="col period-ordering"
  >
    1
  </div>
  <div
    class="col period-id"
  >
    p2
  </div>
  <div
    class="col period-stream"
  >
    Tears of Steel
  </div>
  <div
    class="col period-start"
  >
    <input
      class="form-control "
      disabled=""
      min="00:00:00"
      name="start_2"
      step="1"
      type="time"
    />
  </div>
  <div
    class="col period-duration"
  >
    <input
      class="form-control "
      disabled=""
      min="00:00:00"
      name="duration_2"
      step="1"
      type="time"
    />
  </div>
  <div
    class="col period-tracks"
  >
    <button
      aria-disabled="false"
      class="btn btn-sm m-1 btn-warning"
    >
      0/3 tracks
    </button>
  </div>
</DocumentFragment>
`;
